org.eclipse.buildship.core.test/src/main/groovy/org/eclipse/buildship/core/SynchronizationConcurrencyTest.groovy

```diff
@@ -0,0 +1,96 @@
+package org.eclipse.buildship.core
+
+import org.eclipse.core.resources.IResource
+import org.eclipse.core.runtime.ICoreRunnable
+import org.eclipse.core.runtime.IProgressMonitor
+import org.eclipse.core.runtime.IStatus
+import org.eclipse.core.runtime.NullProgressMonitor
+import org.eclipse.core.runtime.Status
+import org.eclipse.core.runtime.jobs.Job
+
+import org.eclipse.buildship.core.internal.CorePlugin
+import org.eclipse.buildship.core.internal.event.Event
+import org.eclipse.buildship.core.internal.event.EventListener
+import org.eclipse.buildship.core.internal.test.fixtures.ProjectSynchronizationSpecification
+import org.eclipse.buildship.core.internal.workspace.ProjectCreatedEvent
+
+class SynchronizationConcurrencyTest extends ProjectSynchronizationSpecification {
+
+    def "Concurrently executed synchronizations runs sequencially"() {
+        setup:
+        List<Job> syncJobs = (1..5).collect {
+            new SyncJob(dir("location_$it") { file "build.gradle", "Thread.sleep(500)" })
+        }
+        ProjectCreationTracker tracker = new ProjectCreationTracker()
+        CorePlugin.listenerRegistry().addEventListener(tracker)
+
+        when:
+        syncJobs.each { it.schedule() }
+        syncJobs.each { it.join() }
+
+        then:
+        tracker.assertTimeBetweenProjectCreationsHigherThan(500)
+
+        cleanup:
+        CorePlugin.listenerRegistry().removeEventListener(tracker)
+    }
+
+    def "Synchronization requires workspace rule"() {
+        setup:
+        // import a sample project
+        File location = dir('SynchronizationConcurrencyTest')
+        importAndWait(location)
+
+        Job syncJob = new SyncJob(location)
+        when:
+        // start and wait for synchronization with workspace root rule already used
+        Closure synOperation = {
+            syncJob.schedule()
+            assert !syncJob.join(1000, new NullProgressMonitor())
+        }
+        workspace.run(synOperation as ICoreRunnable, workspace.root, IResource.NONE, new NullProgressMonitor())
```

Comment:
⭕️ To avoid the cast to `ICoreRunnable` you could declare it as `ICoreRunnable` rightaway.

Reply:
Done